src/psyclone/psyGen.py

```diff
@@ -6075,46 +6210,53 @@ def declare(self, name, datatype, shape=[], scope='local',
                            If it is an empty list then the symbol represents \
                            a scalar.
-        :param str scope: It is 'local' if the symbol just exists inside the \
-                          kernel scope or 'global_*' if the data survives \
-                          outside of the kernel scope. Note that \
-                          global-scoped symbols also have postfixed \
-                          information about the sharing mechanism, at the \
-                          moment just 'global_argument' is available for \
-                          variables passed in/out of the kernel by argument.
-        :param bool is_input: Whether the symbol represents data that exists \
-                              before the kernel is entered and that is passed \
-                              into the kernel.
-        :param bool is_output: Whether the symbol represents data that is \
-                               survives outside the kernel upon exit.
-        :raises KeyError: The provided name can not be used as key in the
+        :param interface: If not None then indicates that the symbol exists \
+                          outside the local scope and provides the details \
+                          of the sharing mechanism (e.g. routine argument).
+        :type interface: :py:class:`psyclone.psyGen.SymbolInterface` or \
+                         NoneType.
+        :raises KeyError: The provided name cannot be used as a key in the \
                           table.
         '''
         if name in self._symbols:
             raise KeyError("Symbol table already contains a symbol with"
                            " name '{0}'.".format(name))
 
-        self._symbols[name] = Symbol(name, datatype, shape, scope, is_input,
-                                     is_output)
+        self._symbols[name] = Symbol(name, datatype, shape, interface)
 
-    def specify_argument_list(self, argument_name_list):
+    def specify_argument_list(self, argument_name_list, interface=None):
         '''
-        Keep track of the order of the arguments and provide the scope,
-        is_input and is_ouput information if it was not available on the
-        variable declaration.
+        Keep track of the order of the arguments and update the interface
+        of the associated Symbols if we weren't previously able to determine
+        their scope.
 
         :param list argument_name_list: Ordered list of the argument names.
+        :param interface: Interface describing how the supplied arguments \
+                          are accessed in the kernel or None.
+        :type interface: :py:class:`SymbolInterface` or NoneType.
+
+        :raises InternalError: if interface is supplied but is not a \
+                               SymbolInterface.
+
         '''
+        if interface and not isinstance(interface, SymbolInterface):
+            raise InternalError(
```

Comment:
Would `TypeError` fit better?

Reply:
You can probably ignore this one depending how you chose to solve the issue below.